### DISCUSSION TITLE
Correct reference to WASP libraries

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#26557] [Q&A General] [answered]
Hi,
I have some difficulties linking WASP to MOOSE as the default libraries pulled out from 'setup_libtorch.sh' does not seem to link when building moose. 
The libraries from framework/contrib/wasp/install/lib are listed here.
![image](https://github.com/idaholab/moose/assets/40111479/fc4c8c93-7c01-4531-bc52-522a2cf7e30b)
Possibly some of these might need renamed to 'libwasp.so' for moose to interpret correctly at compile time. I tried doing this for libwaspcore.so.4 and it worked. Wanted to know from you the right way to get these libraries to be referenced from moose.
Thanks and regards,
Arun

**Comment by GiudGiud** [comment]
@brandonlangley

**Comment by brandonlangley** [comment]
@abarun22 -

Is there a reason you are not using the recommended Conda environment route for the WASP libraries?

Using the Conda route, these WASP shared libraries should exist in `$WASP_DIR/lib` on your Linux box:

```shell
libwaspcore.so -> libwaspcore.so.4.0.3
libwaspcore.so.04 -> libwaspcore.so.4.0.3
libwaspcore.so.4.0.3
libwaspddi.so -> libwaspddi.so.4.0.3
libwaspddi.so.04 -> libwaspddi.so.4.0.3
libwaspddi.so.4.0.3
libwaspexpr.so -> libwaspexpr.so.4.0.3
libwaspexpr.so.04 -> libwaspexpr.so.4.0.3
libwaspexpr.so.4.0.3
libwasphalite.so -> libwasphalite.so.4.0.3
libwasphalite.so.04 -> libwasphalite.so.4.0.3
libwasphalite.so.4.0.3
libwasphit.so -> libwasphit.so.4.0.3
libwasphit.so.04 -> libwasphit.so.4.0.3
libwasphit.so.4.0.3
libwasphive.so -> libwasphive.so.4.0.3
libwasphive.so.04 -> libwasphive.so.4.0.3
libwasphive.so.4.0.3
libwaspjson.so -> libwaspjson.so.4.0.3
libwaspjson.so.04 -> libwaspjson.so.4.0.3
libwaspjson.so.4.0.3
libwasplsp.so -> libwasplsp.so.4.0.3
libwasplsp.so.04 -> libwasplsp.so.4.0.3
libwasplsp.so.4.0.3
libwaspsiren.so -> libwaspsiren.so.4.0.3
libwaspsiren.so.04 -> libwaspsiren.so.4.0.3
libwaspsiren.so.4.0.3
libwaspson.so -> libwaspson.so.4.0.3
libwaspson.so.04 -> libwaspson.so.4.0.3
libwaspson.so.4.0.3
```

You see that these `libwasp*.so` symbolic links are in the WASP libraries from the Conda environment:

```shell
libwaspcore.so -> libwaspcore.so.4.0.3
libwaspddi.so -> libwaspddi.so.4.0.3
libwaspexpr.so -> libwaspexpr.so.4.0.3
libwasphalite.so -> libwasphalite.so.4.0.3
libwasphit.so -> libwasphit.so.4.0.3
libwasphive.so -> libwasphive.so.4.0.3
libwaspjson.so -> libwaspjson.so.4.0.3
libwasplsp.so -> libwasplsp.so.4.0.3
libwaspsiren.so -> libwaspsiren.so.4.0.3
libwaspson.so -> libwaspson.so.4.0.3
```

So unless you using HPC and cannot go this route, I recommend using Conda as described here for WASP:

https://mooseframework.inl.gov/getting_started/installation/conda.html

If you cannot use Conda and need to manually install WASP, then those `libwasp*.so` links should still exist.

I know nothing about `setup_libtorch.sh`, but it shouldn't have any bearing on the installed WASP libraries.

Again if you can't use WASP from the Conda environment and need to do an offline install, could you try to:

 1. remove the contents of the `framework/contrib/wasp` directory
 2. run the `./scripts/update_and_rebuild_wasp.sh` script
 3. use `ls -ltr framework/contrib/wasp/install/lib` to see if those `libwasp*.so` links were installed

**Comment by abarun22** [comment]
HI @brandonlangley ,
The coda way of installing moose and its dependencies does not come with out any difficulties, as hearing from other users, it is may not be the nicest way dealing with build stuffs and so i have never given a thought about it. 
On the wasp building, that script i was mentioning does set up the libtorch components and not the wasp as such. Apologies for that, as i got mixed-up between these build errors in my moose environment. 

The wasp build script, as i can see, creates the library components along with the symbolic links needed for the linker. However in my case it sill does not reference these components installed in 'framework/contrib/wasp/install/lib' throwing the following error. FYI, this path is referenced in both WASP_DIR  and LD_LIBRARY_PATH for moose to be able to pick at the link time.

```
[ir-bala2@login-q-2 mooseApp]$ make -j10

***ERROR***
WASP does not seem to be available.
Make sure to either run scripts/update_and_rebuild_wasp.sh in your MOOSE directory,
or set WASP_DIR to a valid WASP install
make: *** [/home/ir-bala2/moose_dev/moose/framework/moose.mk:375: wasp_submodule_status] Error 1

```
I do not see any issues with the components either as these were correctly linked to their symbolic names.

![image](https://github.com/idaholab/moose/assets/40111479/8d82c20b-8267-42d4-84b5-da8204dba27a)

You might've better suggestions for fixing this issue that seems quite straight-forward.
Kind regards,
Arun

**Comment by loganharbour** [comment]
You might have a goofed install. Try the following (from the moose root):

```
cd framework
git clean -xfdq
cd ..
git submodule deinit -f framework/contrib/wasp
MOOSE_JOBS=10 ./scripts/update_and_rebuild_wasp.sh
cd test
make -j 10
```

This'll clean up everything in the framework

**Comment by brandonlangley** [comment, marked as answer]
@abarun22 -

Actually, this looks problematic to me:

```
$> echo $WASP_DIR
/home/ir-bala2/moose dev/moose/framework/contrib/wasp/install/lib
```

The `WASP_DIR` variable is supposed to point to `$(MOOSE_DIR)/framework/contrib/wasp/install`.

In your case this would be `/home/ir-bala2/moose dev/moose/framework/contrib/wasp/install` without the `lib` subdirectory.

How / where is this `WASP_DIR` environment variable being set for you? Are you setting that yourself somehow?

You should not have to set that manually since your WASP install is in the normal place from `update_and_rebuild_wasp.sh`.

Because this line in `moose.mk` will set it correctly to that path (without the `lib` subdirectory) if it is not already set:

```
WASP_DIR            ?= $(MOOSE_DIR)/framework/contrib/wasp/install
```

**Reply by abarun22** [reply]
That's pretty much the issue and sorted now.
Thanks @brandonlangley and @loganharbour

**Reply by abarun22** [reply]
I was setting that manually as i could not test it with the wasp install script, which might require a same 'terminal' building of moose. Good to check that although.
